src/shared/bootspec.c

```diff
@@ -757,92 +757,36 @@ static int find_sections(
-        if ((size_t) n != n_sections * sizeof(struct PeSectionHeader))
-                return log_warning_errno(SYNTHETIC_ERRNO(EIO), "%s: Short read while reading sections, ignoring.", path);
-
-        for (size_t i = 0; i < n_sections; i++) {
-                _cleanup_free_ char *k = NULL;
-                uint32_t offset, size;
-                char **b;
-
-                if (strneq((char*) sections[i].Name, ".osrel", sizeof(sections[i].Name)))
-                        b = &osrelease;
-                else if (strneq((char*) sections[i].Name, ".cmdline", sizeof(sections[i].Name)))
-                        b = &cmdline;
-                else
-                        continue;
-
-                if (*b)
-                        return log_warning_errno(SYNTHETIC_ERRNO(EBADMSG), "%s: Duplicate section %s, ignoring.", path, sections[i].Name);
-
-                offset = unaligned_read_le32(&sections[i].PointerToRawData);
-                size = unaligned_read_le32(&sections[i].VirtualSize);
+        _cleanup_free_ IMAGE_SECTION_HEADER *sections = NULL;
+        _cleanup_free_ IMAGE_DOS_HEADER *dos_header = NULL;
+        _cleanup_free_ char *osrel = NULL, *cmdline = NULL;
+        _cleanup_free_ PeHeader *pe_header = NULL;
+        int r;
 
-                if (size > PE_SECTION_SIZE_MAX)
-                        return log_warning_errno(SYNTHETIC_ERRNO(EBADMSG), "%s: Section %s too large, ignoring.", path, sections[i].Name);
+        assert(fd >= 0);
+        assert(path);
 
-                k = new(char, size+1);
-                if (!k)
-                        return log_oom();
+        r = pe_load_headers(fd, &dos_header, &pe_header);
+        if (r < 0)
+                return log_warning_errno(r, "Failed to parse PE file '%s': %m", path);
 
-                n = pread(fd, k, size, offset);
-                if (n < 0)
-                        return log_warning_errno(errno, "%s: Failed to read section payload, ignoring: %m", path);
-                if ((size_t) n != size)
-                        return log_warning_errno(SYNTHETIC_ERRNO(EIO), "%s: Short read while reading section payload, ignoring:", path);
+        r = pe_load_sections(fd, dos_header, pe_header, &sections);
+        if (r < 0)
+                return log_warning_errno(r, "Failed to parse PE sections of '%s': %m", path);
 
-                /* Allow one trailing NUL byte, but nothing more. */
-                if (size > 0 && memchr(k, 0, size - 1))
-                        return log_warning_errno(SYNTHETIC_ERRNO(EBADMSG), "%s: Section contains embedded NUL byte, ignoring.", path);
+        if (!pe_is_uki(pe_header, sections))
+                return log_warning_errno(SYNTHETIC_ERRNO(EBADMSG), "Parsed PE file '%s' is not a UKI.", path);
 
-                k[size] = 0;
-                *b = TAKE_PTR(k);
-        }
+        r = pe_read_section_data(fd, pe_header, sections, ".osrel", PE_SECTION_SIZE_MAX, (void**) &osrel, NULL);
```

Comment:
also here, osrel can be missing as per spec

Reply:
we did the pe_is_uki() check first, so this cannot be missing...

we can of course change the pe_is_uki() check, as discussed above, but let's look into this in a separate PR, as mentioned.